### PR TITLE
Running prettier on commit

### PR DESCRIPTION
**Pull request by olemartinorg**
This conflicts with #77. Adds [prettier](https://prettier.io/) with [husky](https://github.com/typicode/husky) and [lint-staged](https://github.com/okonet/lint-staged) in order to enforce code style and automatically reformat on commit. It also adds [eslint-plugin-prettier](https://github.com/prettier/eslint-plugin-prettier) in order to run prettier on `eslint --fix` (like we have now), but also have eslint report prettier conflicts as errors.

The plan is to run prettier over the entire code base once after this (will do that in a separate PR).

### CODE REVIEW

**Comment by olemartinorg** [comment]
Well, eslint is NOT happy about this.. Probably should have expected this. What to do? Should I run the entire codebase through prettier as a last huge commit to add to this PR, instead of doing it in a later PR?

File: src/package.json
Comment (haakemon) [inline src/package.json:21]: I would prefer to use 
```suggestion
    "husky": "^6",
    "lint-staged": "^10",
```
to avoid potentially installing a major version with breaking changes.

File: src/package.json
Comment (haakemon) [inline src/package.json:31]: Should we add an entry for json files as well?

**Review by haakemon — APPROVED** [review]

**Comment by haakemon** [comment]
> Well, eslint is NOT happy about this.. Probably should have expected this. What to do? Should I run the entire codebase through prettier as a last huge commit to add to this PR, instead of doing it in a later PR?

Would be good to just do it in one go I think, but fixing the issues will cause a lot of noise for all the open PR's wont it? Should we perhaps wait to fix & merge this until most of the other PR's are merged?

**Comment by olemartinorg** [comment]
> fixing the issues will cause a lot of noise for all the open PR's wont it?

Yes and no. I'm not 100% certain here, but I believe it could cause less noise, although it would probably also cause some (easy to fix) merge conflicts.

Let's take the Likert component PR; it's a large change, and parts of the patches contain code style changes only. If we run prettier on the main branch before merging the PR, git should be smart enough to notice that the result of the patch is the same as in main, so there should not be any merge conflicts. The result for these patch segments is less noise in the PR, because pure code style changes from the PR will be removed from review.

However, when there are patch segments that contain both code style changes **and** other changes, the patch source now differs from main, and you'll get merge conflicts. However, although tedious, these should be easy to resolve (just overwrite what's on main).

The easiest for everyone might still be to wait until we have only a few/no other PRs, if that is feasible.

**Review by mjulstein — APPROVED** [review]
I say merge it asap!
However maybe after the version-tag-bump-thingy?
Do we need prettier formatting validation as a part of the tests when we are going to format on commit anyway?

**Comment by olemartinorg** [comment]
I suggest waiting until a few more of the larger PRs have been merged, especially if we're going to run prettier on the entire code base before merging this. I had some practice yesterday, as the same change was made to main that I had also done in my private branch. While merging main into my branch, the whole file showed up as a conflict because there was a difference in a single import location in the top of the file. That kind of hassle is not worth it for these large PRs in our queue, I think. I suggest we hold on a bit more and coordinate with the team to find a good day to do a sweeping change like this.

In the meantime I suggest we get #77 merged.

File: src/package.json
Comment (olemartinorg) [reply src/package.json:17]: Huh.. I guess I just followed the example code from @nemzes on Slack. With that said, it's not surprising the prettier team is preaching "don't wrap me". :wink: I guess we could run prettier first and eslint second, but I fear it will complicate things if we require them to run in a specific order instead of just running one tool where we can configure the order. I checked now, and while I haven't tested it to make sure, it seems if you configure WebStorm to run both eslint and prettier on save, it runs eslint first and prettier second - which would conflict with having something else happen automatically on commit.

Or, I may just have assumed we'll want to run both eslint and prettier on commit - but maybe we just want one of them? I believe we can use [import-js/eslint-plugin-import](https://github.com/import-js/eslint-plugin-import) to automatically use `import type` and group these type imports separately. I've seen that being a common feedback in code review, so it would be nice to just automate it for us. For that we'd want to run both tools on commit.

Comment (haakemon) [reply src/package.json:17]: Not sure what is best :) I just thought since its not recommended, maybe its good to follow that. But if it complicates other things, it may not be worth to follow those recommendations.

+1 for using [import-js/eslint-plugin-import](https://github.com/import-js/eslint-plugin-import) - I didnt know about that one. Seems like a useful addition.

Comment (haakemon) [reply src/package.json:17]: Lets leave it in for now, and revisit if it actually causes issues :)

Comment (olemartinorg) [reply src/package.json:17]: Agree! We'll quickly notice if it's too slow or buggy.

**Comment by haakemon** [comment]
I'm wondering if we should add `singleAttributePerLine` option that was [released in prettier 2.6](https://prettier.io/blog/2022/03/16/2.6.0.html) as well? Personally I've been wanting this feature for a long time, and it avoids getting the unnescessary diffs when adding that one extra prop that makes the line break onto separate lines, since they now always would be on separate lines.

**Comment by olemartinorg** [comment]
I agree, @haakemon! Added your suggestion now, along with the `import type` eslint rule we've talked about.

When merging in main, the git hooks ran on the merge commit, so now I've successfully tested prettier on commit.. :partying\_face: Sorry for all the review noise, but there's no point cleaning it up if this pull request will include a full run on the entire code base anyway - probably best to just review the merge commit on its own (with the minor conflict resolutions in `package.json` and `yarn.lock`).

File: src/altinn-app-frontend/.prettierrc.json
Comment (haakemon) [inline src/altinn-app-frontend/.prettierrc.json:2]: @olemartinorg I missed this during review, there is a prettier config on root which is the one that should be used. Now we have a separate config for altinn-app-frontend subfolder which only has this rule (and prettier default for rest) and other rules (root folder config) for all other files...

Comment (haakemon) [reply src/altinn-app-frontend/.prettierrc.json:2]: Fixed in https://github.com/Altinn/app-frontend-react/pull/302 🙂

**Pull request merged**
